File: baekjoon/dp/java/BOJ11066.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class BOJ11066 {
    private static BufferedReader br;
    public static void main(String[] args) throws NumberFormatException, IOException {
        br = new BufferedReader(new InputStreamReader(System.in));       
        
        int T = Integer.parseInt(br.readLine());

        for (int i=0;i<T;i++) {
            System.out.println(mergeFiles());
        }
    }

    public static int mergeFiles() throws NumberFormatException, IOException {
        int N = Integer.parseInt(br.readLine());
        int[] files = getFiles(N);
        int[] acc = accumulate(files);
        int[][] dp = new int[N][N];

        for (int step=1;step<N;step++) {
            for(int start=0;start<N-step;start++) {
                int end = start + step;

                dp[start][end] = Integer.MAX_VALUE;

                for (int j=start;j<end;j++) {
                    dp[start][end] = Math.min(
                        dp[start][end],
                        dp[start][j] + dp[j+1][end] + acc[end+1] - acc[start]);
                }
            }
        }

        return dp[0][N-1];
    }

    public static int[] accumulate(int[] arr) {
        int[] res = new int[arr.length+1];

        for (int i=0;i<arr.length;i++) {
            res[i+1] = res[i] + arr[i];
        }
        return res;
    }

    public static int[] getFiles(int N) throws IOException {
        int[] files = new int[N];
        StringTokenizer st = new StringTokenizer(br.readLine(), " ");
        
        for (int i=0;i<N;i++) {
            files[i] = Integer.parseInt(st.nextToken());
        }
        return files;
    }
}

File: baekjoon/dp/python/BOJ11066.py
from sys import stdin
from collections import defaultdict

readline = stdin.readline

T = int(readline())

def solve():
    for _ in range(T):
        print(merge_files())

def merge_files():
    N = int(readline())
    files = list(map(int, readline().split()))

    acc = accumulate(files)
    dp = [[0 for _ in range(N)] for _ in range(N)]

    for step in range(1, N):
        for start in range(N-step):
            end = start + step
            dp[start][end] = float('inf')
            for j in range(start, end):
                dp[start][end] = min(
                    dp[start][end], 
                    dp[start][j] + dp[j+1][end] + acc[end] - acc[start-1])
    
    return dp[0][N-1]

def accumulate(arr):
    res = defaultdict(int)
    for i in range(len(arr)):
        res[i] = res[i-1] + arr[i]

    return res

solve()